servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -0,0 +1,1063 @@
+import tachyon.master.file.meta.InodeDirectoryIdGenerator;
+import tachyon.master.file.meta.InodeFile;
+import tachyon.master.file.meta.InodeTree;
+import tachyon.master.journal.Journal;
+import tachyon.master.journal.JournalEntry;
+import tachyon.master.journal.JournalOutputStream;
+import tachyon.thrift.BlockInfo;
+import tachyon.thrift.BlockInfoException;
+import tachyon.thrift.BlockLocation;
+import tachyon.thrift.DependencyDoesNotExistException;
+import tachyon.thrift.DependencyInfo;
+import tachyon.thrift.FileAlreadyExistException;
+import tachyon.thrift.FileBlockInfo;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.FileInfo;
+import tachyon.thrift.FileSystemMasterService;
+import tachyon.thrift.InvalidPathException;
+import tachyon.thrift.NetAddress;
+import tachyon.thrift.SuspectedFileSizeException;
+import tachyon.thrift.TachyonException;
+import tachyon.underfs.UnderFileSystem;
+import tachyon.util.FormatUtils;
+import tachyon.util.ThreadFactoryUtils;
+import tachyon.util.io.PathUtils;
+
+public final class FileSystemMaster extends MasterBase {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final TachyonConf mTachyonConf;
+  private final BlockMaster mBlockMaster;
+
+  /** This manages the file system inode structure. This must be journaled. */
+  private final InodeTree mInodeTree;
+  /** This manages metadata for lineage. This must be journaled. */
+  private final DependencyMap mDependencyMap = new DependencyMap();
+  /** This generates unique directory ids. This must be journaled. */
+  private final InodeDirectoryIdGenerator mDirectoryIdGenerator;
+
+  private final PrefixList mWhitelist;
+
+  /**
+   * @param baseDirectory the base journal directory
+   * @return the journal directory for this master
+   */
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.FILE_SYSTEM_MASTER_SERVICE_NAME);
+  }
+
+  public FileSystemMaster(TachyonConf tachyonConf, BlockMaster blockMaster,
+      Journal journal) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("file-system-master-%d", true)));
+    mTachyonConf = tachyonConf;
+    mBlockMaster = blockMaster;
+
+    mDirectoryIdGenerator = new InodeDirectoryIdGenerator(mBlockMaster);
+    mInodeTree = new InodeTree(mBlockMaster, mDirectoryIdGenerator);
+
+    // TODO: handle default config value for whitelist.
+    mWhitelist = new PrefixList(mTachyonConf.getList(Constants.MASTER_WHITELIST, ","));
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new FileSystemMasterService.Processor<FileSystemMasterServiceHandler>(
+        new FileSystemMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getServiceName() {
+    return Constants.FILE_SYSTEM_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
```

Comment:
as per my comment in the BlockMasterClient it would be better to use polymorphism here